pymbolic/mapper/substitutor.py

```diff
@@ -39,14 +46,21 @@ def map_subscript(self, expr):
         if result is not None:
             return result
         else:
-            return pymbolic.mapper.IdentityMapper.map_subscript(self, expr)
+            return IdentityMapper.map_subscript(self, expr)
 
     def map_lookup(self, expr):
         result = self.subst_func(expr)
         if result is not None:
             return result
         else:
-            return pymbolic.mapper.IdentityMapper.map_lookup(self, expr)
+            return IdentityMapper.map_lookup(self, expr)
+
+
+class CachedSubstitutionMapper(CachedIdentityMapper,
+                               SubstitutionMapper):
+    def __init__(self, subst_func):
+        CachedIdentityMapper.__init__(self)
+        SubstitutionMapper.__init__(self, subst_func)
```

Comment:
Does this actually return from the cache for map_variable, map_subscript, map_lookup?

Reply:
Why wouldn't it?

Reply:
I was just wondering because the SubstitutionMapper from which this is derived directly calls the IdentityMapper for these methods. But if I understood the MRO here correctly, this appears to not be an issue.

Reply:
> SubstitutionMapper from which this is derived directly calls the IdentityMapper for these methods.

Cache lookups happen in `rec`, but `SubstitutionMapper` doesn't override `rec`. I think this is OK, but maybe I'm missing something.